### DISCUSSION TITLE
Micro-frontend fragment architecture

### DISCUSSION THREAD

**Discussion by petebacondarwin** [cloudflare/workers-web-experiments#65] [General]
What do you think about creating collaborating server-side rendered micro-frontend fragments with Cloudflare Workers, as shown in our cloud-gallery demo?

**Comment by luisvieiragmr** [comment]
i still think there are some issues when loading multiple fragments in the same page especially for bigger and complex pages where multiple fragments may use the same components. 
Despite using a low overhead framework such as Qwik if multiple fragments depend on a complex and large lib for some functionality the overhead will start to come up. 
Also multiple fragments may have similar data needs and therefore i think there can be some extra http requests for data fetching in this type of architecture

**Reply by petebacondarwin** [reply]
These are very valid points. As soon as there are common aspects between fragments then there is likely to be duplication. This is one of the trade-offs that must be considered. Do you have ideas how we might mitigate some of these problems that you have identified?

One of the benefits of SSR+Qwik is that the JS that needs to be downloaded is not proportional to the size of the application (as it is for just about all the other frameworks). Instead it is related to the parts of the app that are interacted with. In this case, you only pay for (in terms of JS) for components that need to be re-rendered. You might have numerous fragments that all contain a complex datepicker. But if the user only interacts with one of them, only one copy of the datepicker code would need to be downloaded and executed.

Related to this, whatever approach you use to create micro-frontends, you might still come unstuck if there are different versions involved. If there are multiple fragments using different versions of the aforementioned datepicker, then you have no choice but to download them all eventually. This is compounded if the fragments use different frameworks or versions thereof.

Regarding data duplication. We are looking at how some meta-frameworks (e.g. QwikCity or Remix) provide standardized data access strategies where you can narrowly define and provide only the data that is specifically needed for the current fragment. Often, with REST, one can end up providing the whole world, since it is not so easy to distinguish the data required for a particular part of the app. Things like GraphQL can help with this.

With SSR and frameworks like Qwik you can often avoid making client requests for data until absolutely necessary, which can also help with this problem. Further with Cloudflare Workers judicious use of data/request caching could also mean that any requests from the client (for data, or even SSRed HTML) are returned directly from the Worker in a very short time, given the low latency of Workers.

**Reply by luisvieiragmr** [reply]
i think in order to address the duplication issue there are no solutions without increasing coupling between fragments.
I've been thinking about a fragment that provides some common dependencies however this solution comes with lots of trade-offs and complexity, also i'm still not sure about the webpack module federation approach, it think it can have some resiliency issues, but something i'll explore in the future.

My data fetching concerns were more around in a page composed of multiple fragments sometimes fragments have similar data needs, for example in a e-commerce page multiple fragments can show the same product and check if it's on a wishlist or bag, even with frameworks like remix where you can defer this data fetching to the server you'll still end up with extra load in your api's and therefore incur in extra costs and scaling needs, so i think it's a trade-off to consider.

Also when transitioning from page to page with a SPA style navigation you may want to use client cached data to display something to the user immediately while fetching missing data, to enable something like this probably fragments would need to be aligned on a api client with common caching between fragments

Another issue is what kinds of operations are allowed in each fragment backend because if they start to add some specific business logic then it's easy to end up with inconsistent data in the application

**Comment by Abiti-936** [comment]
I think these is just a super cool architecture for managing dynamic applications at scale, on the edge (closer to users).

I have also tried these architecture for a Rest API & It's just very very much great, as 

- these communication between these bound workers is instant (as they are running in a same thread)
- we can structure a project in a mono repo tool like [Turbo Repo](https://turbo.build/repo) or with multiple Repos
- we can deploy and iterate independently & efficiently
- local development is just really seamless with Wrangler CLI

Also, How is these architecture is going to integrate with Cloudflare Products like KV, D1 Database, Durable Object, R2 Storage, Queues & Pages Functions?

So, Keep working on these great architecture. Thanks.
